silkworm/silkrpc/core/fee_history_oracle.cpp

```diff
@@ -0,0 +1,138 @@
+        if (reward_percentile[idx] < 0 || reward_percentile[idx] > 100) {
+            std::ostringstream ss;
+            ss << "ErrInvalidPercentile: " << std::dec << reward_percentile[idx];
+
+            fee_history.error = ss.str();
+            co_return fee_history;
+        }
+        if (idx > 0 && reward_percentile[idx] < reward_percentile[idx - 1]) {
+            std::ostringstream ss;
+            ss << "ErrInvalidPercentile: #" << idx - 1 << ":" << reward_percentile[idx - 1]
+               << "> #" << idx << ":" << reward_percentile[idx];
+            fee_history.error = ss.str();
+            co_return fee_history;
+        }
+    }
+
+    auto max_history = reward_percentile.size() > 0 ? DEFAULT_MAX_BLOCK_HISTORY : DEFAULT_MAX_HEADER_HISTORY;
+
+    auto block_range = co_await resolve_block_range(newest_block, block_count, max_history);
+
+    fee_history.rewards.reserve(block_range.num_blocks);
+    fee_history.base_fees_per_gas.reserve(block_range.num_blocks + 1);
+    fee_history.gas_used_ratio.reserve(block_range.num_blocks);
+
+    auto oldest_block = block_range.last_block + 1 - block_range.num_blocks;
+    for (auto idx = block_range.num_blocks; idx > 0; idx--) {
+        auto block_number = ++oldest_block - 1;
+
+        if (block_number > block_range.last_block) {
+            continue;
+        }
+        BlockFees block_fees{block_number};
+
+        if (block_number >= block_range.block.block.header.number) {
+            block_fees.block = block_range.block;
+            block_fees.receipts = co_await receipts_provider_(block_fees.block);
+            ;
+        } else {
+            block_fees.block = co_await block_provider_(block_number);
+            if (reward_percentile.size() > 0) {
+                block_fees.receipts = co_await receipts_provider_(block_fees.block);
+            }
+        }
+
+        co_await process_block(block_fees, reward_percentile);
+        auto index = block_fees.block_number - oldest_block;
+        fee_history.rewards[index] = block_fees.rewards;
+        fee_history.base_fees_per_gas[index] = block_fees.base_fee;
+        fee_history.base_fees_per_gas[index + 1] = block_fees.next_base_fee;
+        fee_history.gas_used_ratio[index] = block_fees.gas_used_ratio;
+    }
+    // TODO firstMissing management as in erigon
+    fee_history.oldest_block = newest_block;
+    fee_history.base_fees_per_gas = {0x13c723946e, 0x163fe26534};
+    fee_history.gas_used_ratio = {0.9998838666666666};
+    fee_history.rewards = {Rewards{0x59682f00, 0x9502f900}};
+
+    co_return fee_history;
+}
+
+boost::asio::awaitable<BlockRange> FeeHistoryOracle::resolve_block_range(uint64_t /*last_block*/, uint64_t /*block_count*/, uint64_t /*max_history*/) {
+    /*const auto block_with_hash = co_await block_provider_(last_block);
+    const auto receipts = co_await receipts_provider_(block_with_hash);*/
+
+    BlockRange block_range{0};
+
+    co_return block_range;
+}
```

Comment:
This seems like unfinished work to me or maybe you're just missing one last push :smile: